### ISSUE TITLE
Allow to specify custom name for bicep deployments

### ISSUE THREAD

**Issue by JayDoubleu** [Azure/azure-dev#2320]
Currently, Bicep deployments are created with randomized names. This leads to a high number of deployments at the subscription level.

It would be beneficial to allow users to specify custom deployment names, so that they can be represented by a single deployment run. This feature would also be useful for Bicep deployments that use deployment names to create role assignment GUIDs. If a different role assignment name already exists for the same resource, it could lead to errors.

**Comment by rajeshkamal5050** [comment]
@ellismg @weikanglim do we support/have any environment variables for customizing names?

**Comment by wbreza** [comment]
All of the Bicep modules used in the azd templates support optional parameters that can be used to override the resource names.  Resource group names can also be overridden inside the azure.yaml

**Comment by ellismg** [comment]
> Currently, Bicep deployments are created with randomized names. This leads to a high number of deployments at the subscription level.

This is by design.  We think that out of the box there is a tremendous about amount of value in having a separate deployment object per invocation of `azd provision`, for history and auditing.  Earlier iterations of the tool actually used a fixed deployment name and we explicitly made the change to use a unique deployment name per deployment.


> This feature would also be useful for Bicep deployments that use deployment names to create role assignment GUIDs. If a different role assignment name already exists for the same resource, it could lead to errors.

The names that we generate follow the following pattern:

`<environment-name>-<unix-timestamp-of-when-the-deployment-was-created>`

You should be able to take a dependency on that in bicep use the [string functions](https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/bicep-functions-string) to pull apart the environment name from the timestamp by looking for the last index of the hypen and the substringing everything before that.

> @ellismg @weikanglim do we support/have any environment variables for customizing names?

We don't have any support today for customizing this name.  Unlike what wallace proposes, the name of the root deployment is computed by the cli itself, not a template.  If we wanted to add such a thing, I imagine the way we would want to do it is to add something to the `infra` block of `azure.yaml` that would allow you to provide a template (that we'd do our normal .env-var replacement on) for the deployment name.  If we had this, @JayDoubleu could set this as something like this:

```yaml
infra:
    deploymentName: ${AZURE_ENV_NAME}
```

To use the same deployment name per environment.

**Comment by weikanglim** [comment]
> It would be beneficial to allow users to specify custom deployment names, so that they can be represented by a single deployment run. This feature would also be useful for Bicep deployments that use deployment names to create role assignment GUIDs. If a different role assignment name already exists for the same resource, it could lead to errors.

@JayDoubleu would love to understand the scenario here more and see if there's a different construct `azd` could offer here. Do you mean that:

Given a single "run" (which can contain a series of bicep deployments), we want the multiple deployment instances to share the same name? I wonder if this the logic can all be keyed from a bicep parameter "runId" of the likes, and keeping deployment name unique per instance.

As @ellismg mentions above, there are nice things users do get when having unique deployments. The deployments form audit trails and history. azd also uses deployments to retrieve state, and this allows you essentially "roll-back" to reference a previous state -- although how this can be used in practice is unlikely. ARM deployment history is also automatically [cleaned up](https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/deployment-history-deletions?tabs=azure-powershell) (this didn't use to be the case).

**Comment by JayDoubleu** [comment]
Thanks @weikanglim so the use case for this would be exactly the same one allowed by azure cli. You can define the deployment name with `--name -n` or you can omit it completely and random deployment name would be selected for you.

I would find the custom name particularly useful for subscription scoped deployments where you can have 30 different people running their own deployments multiple times a day and if the names are randomised then it can generate a lot of noise in deployments blade. For that reason it would be great to have an option to manually specify it (or perhaps allow some GO functions to be specified as part of the deployment name so that it will allow the user to specify the randomness of the deployment etc)

**Comment by weikanglim** [comment]
@JayDoubleu Thanks for the write-up!

Our team also has a shared playground subscription where we have devs from various teams all deploying resources. We've created an unspoken best practice to name azd environments with "<username>-<actual env name>-<unique suffix>" as a way to enforce uniqueness, and so far it's worked well. I suppose that since deployment name derives from environment name, you could use this for your own benefit as well.

**Comment by JayDoubleu** [comment]
I agree. This is customisable. Agreeing or enforcing the opinionated naming conventions in large organisation is another challenge. The "shared playground subscription" is not the same as multi hub/spoke-multi-tier/multi-team/multi-project organisation and things can get out of hand quickly. 

In my opinion it's better to offer the same choice bicep and azure cli offers
